PR Task 002

SergioLuis opened the pull request:
The general execution infrastructure is in place. The code is pretty self-explanatory, but requires the reader to be familiar with how the original system worked. You can check out the online available documentation.

Right now, the emulator is able to load a ROM and start the execution (single-threaded). The `JP addr`, `CLS` and `LD Vx, Byte` instructions are already implemented and tested as an example on how to do so.

@jhvargas3112 @ugedo please get familiar with the code and ask any doubts you have. I'll divide the instruction set soon and everybody will get to work on this core.

I created the Pull Request in order to get a formal review on the code. Some work is still to be done (for example, loading an snapshot), but I think we should leave that for later, as this task has gotten out of hand in invested hours terms.

mikelcaz commented:
- 927dd0f Change the PC to Int to avoid conversions

Mmm... i'm not very sure about this. We do not the same with memory (cause IMemory<whatever>), but it's possible expect an IMemory<Int> where we normally expect an IMemory<Byte>.

It's complicated. If we cannot agree here, we can just use this.

- a4b006b Add a 'PaddedMemory' to emulate original memory
- 5386818 Add an intermediate LoggedMemory

I agree with using this aproach. Note that we used specialized classes like ByteArray and ShortArray for the non-padded classes. I like using a generic, but I'm not sure about the pros of using these specialized classes. **So we should discuse what is better and stick with it**.

- 77f1b5d Refactor memory classes to receive an array

Good, very flexible. But crap. Broken (*nix-like) lines are creeping into the code. **All the files without a newline character at the end must be fixed before merging this into dev**.

- 9f994ab Define the IRenderer interface

Also:  
- 375285f Add method to the IRenderer to clear the display

Related to:
- 322d564 Add more members to the Cpu

A (so called) present method should be added, or the interface renamed. A renderer is not a window/screen/surface. Is a renderer. Clear should be called ever before calling any rendering functions (only the render method in this case), and present after all that to dump. This allows implementing things like double buffering, vsync and such.

But I'm not sure about this. Maybe we don't need a renderer interface in first place, so **we should discuss this topic before the merge**.

- 758ee55 Minor refactor to IClockDivider and subtypes

> The function 'msLeft' should be called 'getMsLeft'. Also, I don't
> feel comfortable initializing the state of the ClockDivider directly
> on the constructor, as it might not be used until after a while of
> its creation.

- First, getMsLeft should be a **read-only property called msLeft** (my fault, sorry).
- Second, I think init should be called reset. There is no quit, and must be callable after the first time (for example after resuming).
- Third, should be stated if calling init/reset before using is recommended or required. Note that in any case, initializing to zero in the constructor backfires under both criteria.

- 322d564 Add more members to the Cpu
- e51b0aa Add an interface to sleep the thread

Makes sense. Seems ok to me, but I would prefer to rename milliseconds as _ms_ for consistency.

- a8d0d49 Start implementing the Chipeit class

Also:  
- 6e98d44 Fix the padding while creating the memory

Related to:  
- 7f0539f Remove unused method on IClock interface

It looks good! That eventTimer should be _delayTimer_. Yes, we need clarify its purpose. Core imports are redundant in a contradictory style (f*cking IntelliJ). **And the clock should be updated**.

- 7f0539f Remove unused method on IClock interface

Also ():  
- 7e47679 Implement same-thread "dummy" code execution

**No way**. This should be undone. Every element in the system is sincronized with the same clock. Update makes and snapshot.

I think we should replace getMs with a read-only property (clock.ms which I prefer, or clock.timestamp. My fault again).

- 66b7649 Simplify instruction reading

But it is not simpler!

- 4d69992 Jump instruction implemented and tested!

Seems ok. But I'm not very sure about what the caller should expect from jpAddr, and specially if the caller name es decodeAndExecute.

- 9815f38 Implement and test the ldVxByte instruction

Seems ok (but byte var at ldVxByte should be renamed).

- b3d063e Work on the instructions that start with zero

Seems ok (I'm surprised about modern implementations ignore _SYS addr_, I wonder what should happen next).

SergioLuis commented:
First, my responses to some of the comments:

* About 927dd0f (Change the PC to Int to avoid conversions)

IMemory<T> uses an Integer as the index of the get and set operations.

* About a4b006b (Add a 'PaddedMemory' to emulate original memory)

The `PaddedMemory` class is only to avoid other less elegant solutions such as changing the PC on each call. The specialized classes are there to take advantage of `ByteArray` and `ShortArray`, which are converted to `byte[]` and `short[]` instad to a generic `Array<Byte>` or `Array<Short>`. If we didn't use specialized classes, we could get rid of ByteMemory and ArrayMemory and just use a generic `Memory<T> : IMemory<T>`. Because we rely on the interface instead of on the implementation, we can later change this effortlessly.

* About 77f1b5d (Refactor memory classes to receive an array)

Fixed the EOF of IMemory.kt and README.md at bc87cce

* About 9f994ab (Define the IRenderer interface) and 375285f (Add method to the IRenderer to clear the display)

> A renderer is not a window/screen/surface. Is a renderer.

Correct.

> Clear should be called ever before calling any rendering functions.

Incorrect. There is an specific instruction to clear the display, is none of the emulator bussiness.

> This allows implementing things like double buffering, vsync and such.

That is none of the emulator bussiness. The client can implement a double-buffer just with the `render` method.

> Maybe we don't need a renderer interface in first place

What do you suggest to issue draw commands in a platform-agnostic way?

* About 758ee55 (Minor refactor to IClockDivider and subtypes)

Converted the function `getMsLeft` to the read-only property `msLeft` at 81d0c04.  
Renamed the function `init` to `reset` at eb56498

About this:

> ...should be stated if calling init/reset before using the object is recommended or required.

I think it should be REQUIRED before entering the trigger-sleep while-true loop. BUT, as it is an internal part, I don't want to add a way to enforce it.

* About e51b0aa (Add an interface to sleep the thread)

I would prefer to rename every `ms` appearance to `milliseconds` (for consistency), but I complied with the review comment at 7c3cd2b

* About 7f0539f (Remove unused method on IClock interface)

Applied review comments at 76431e2

* About 66b7649 (Simplify instruction reading)

Simplified instruction reading at 9b9c4d8

* About 9815f38 (Implement and test the ldVxByte instruction)

Renamed the `byte` val to `b` at a1b6015
------

Also:

* 927dd0f Delete MyClass.kt example file

mikelcaz commented:
The most important stuff:

- About a4b006b (Add a 'PaddedMemory' to emulate original memory)

Yes, of course, but the point is we are using two different criterion on two similar things without justification.

**Performance vs flexibility**. At first glance we could end using just IMemory<Byte> implementations. I don't care which of the two will be picked. I'm just saying that we should pick the same in both places.

On a bigger system probably both would be implemented following the Jetbrains criterion. It's not a bigger system. Pick one. You choose. It will be fine.

- About 9f994ab (Define the IRenderer interface) and 375285f (Add method to the IRenderer to clear the display)

> Incorrect. There is an specific instruction to clear the display, is none of the emulator bussiness.

And:

> That is none of the emulator bussiness. The client can implement a double-buffer just with the render method.

My fault again. My wording was confusing.

I talked about a renaming OR making the renderer... a renderer. That lines were about the second option. In my opinion a renderer suggest indirection, a intermediary that "renders" something on the surface of another thing (eg. a window, a texture, whatever).

That way, "clear" means something different from that would be expected from, let's call it, a _surface_. Moreover, a renderer can be redirected to another target on the go.

Our interface is different. It relies on a single posible operation that **blits** the whole screen. If multiple operations were allowed, you couldn't implement double buffer or vsync (cause that would be unknown when the current frame is ready). In any case, is more like a surface.

> What do you suggest to issue draw commands in a platform-agnostic way?

I have an idea, but I need further information about what would be handy and the underlying system (yes, besides the platform-agnostic part, don't worry). I'm on the way in this one...

- About 758ee55 (Minor refactor to IClockDivider and subtypes)

> ...should be stated if calling init/reset before using the object is recommended or required.

And:

> [...] I don't want to add a way to enforce it.

Yes, not at the interface. I agree. But read my note again.

1. It's _required_? Take measures at the implementation. Don't set lastTime to 0.
2. It's _recommended_. Well, ok. Don't set lastTime to 0.

But I just see now a new and more important issue: **reset is not even needed for resuming!**

Implementing stop/resume must be done at the clock. Moreover: **using reset instead over different divider objects makes them out of sync**.

Remember, that's the purpose of using a Clock: it's our point of support, as it only "tells the time".

- About 9815f38 (Implement and test the ldVxByte instruction)

> Renamed the byte val to b at a1b6015

What th...?! Ok, my fault again. I did not explain the problem before. The variable was called byte (what we want to load? Bitch please, a byte), which is fairly logical... but it was an Int. Later, that variable has to be converted to Byte (with the toByte method). Which is damn confusing.

Calling it just "b" is that kind of thing you dislike (AND it does nothing). But we have two better alternatives:

1. Call it "value". It has an advange: refactoring several instructions will be easier in the future.
2. **Keep the previous name** (byte), but make the toByte at the same line.
